scripts/release_gen.py

```diff
@@ -0,0 +1,512 @@
+#!/usr/bin/env python3
+import argparse
+import json
+import re
+import shlex
+import subprocess
+import sys
+import time
+from datetime import datetime, timezone
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List, Optional, Sequence, Tuple
+
+
+def header(msg: str) -> None:
+    print(f"==> {msg}", file=sys.stderr)
+
+
+USAGE_TEXT = """
+Usage: scripts/release_gen.py [--dump-only] [-q|--quiet] [--repo <owner/repo>] [--repo-dir <path>] [--gh-timeout-secs N] [--codex-timeout-secs N] [owner/repo] <from_tag> <to_tag> [version]
+
+Examples:
+  scripts/release_gen.py --repo openai/codex v0.23.0 v0.24.0
+  scripts/release_gen.py --repo-dir ../codex-repo v0.23.0 v0.24.0  # detect repo from this directory's git remote
+  scripts/release_gen.py v0.23.0 v0.24.0                            # auto-detect repo from current dir's git remote
+  scripts/release_gen.py v0.23.0 v0.24.0 0.24.0                     # auto-detect with explicit version
+  scripts/release_gen.py --dump-only v0.23.0 v0.24.0                # only generate releases/release_dump_<ver>.txt
+  scripts/release_gen.py -q v0.23.0 v0.24.0                         # quiet Codex call with progress dots
+
+Notes:
+  - Requires: gh (GitHub CLI) for dump generation; codex CLI for note generation.
+  - If release_dump_<ver>.txt does not exist, it will be created automatically.
+  - Then runs codex to generate <ver>.txt based on the dump (unless --dump-only).
+  - If you omit tags, the script lists the last 20 releases for the repo.
+  - Timeouts: set with --gh-timeout-secs (default 60) and --codex-timeout-secs (default 300).
+""".strip()
+
+
+# -------- argument parsing (shell-like) --------
+
+
+@dataclass
+class Args:
+    dump_only: bool
+    quiet: bool
+    repo: Optional[str]
+    repo_dir: Optional[Path]
```

Comment:
Prefer modern types:

```suggestion
    repo: str | None
    repo_dir: Path | None
```